lib/structuraid_core/design_codes/aci_318_19/rc/footings/punching_critical_section_perimeter.rb

```diff
@@ -0,0 +1,132 @@
+               { from: 3, to: 4 }, { from: 4, to: 1 }]
+            end
+
+            def add_perimeter_vertices_to_local_coordinates_system
+              perimeter_vertices_relative_to_column_location.each do |perimeter_vertex|
+                add_relative_location_from_a_vector(
+                  perimeter_vertex + column_relative_location_vector
+                )
+              end
+            end
+
+            def add_column_to_local_coordinates_system
+              add_relative_location_from_a_vector(
+                column_absolute_location.to_vector - local_coordinates_system.anchor_location.to_vector
+              )
+            end
+
+            def update_location_to_limit(location)
+              location.value_1 = udate_to_constrain(location.value_1, 0.5 * footing.length_1)
+              location.value_2 = udate_to_constrain(location.value_2, 0.5 * footing.length_2)
+            end
+
+            def udate_to_constrain(initial_value, constrain_value)
+              return constrain_value if initial_value > constrain_value
+              return -constrain_value if initial_value < -constrain_value
+
+              initial_value
+            end
+
+            def into_footing?(location)
+              vertical_check(location) && horizontal_check(location)
+            end
+
+            def vertical_check(location)
+              location.value_2 >= -0.5 * footing.length_2 && location.value_2 <= 0.5 * footing.length_2
+            end
+
+            def horizontal_check(location)
+              location.value_1 >= -0.5 * footing.length_1 && location.value_1 <= 0.5 * footing.length_1
+            end
+
+            def perimeter_vertices_relative_to_column_location
+              [
+                perimeter_vertices_relative_to_column_location_with(way_1: 1, way_2: 1),
+                perimeter_vertices_relative_to_column_location_with(way_1: -1, way_2: 1),
+                perimeter_vertices_relative_to_column_location_with(way_1: -1, way_2: -1),
+                perimeter_vertices_relative_to_column_location_with(way_1: 1, way_2: -1)
+              ]
+            end
+
+            def perimeter_vertices_relative_to_column_location_with(way_1:, way_2:)
+              Vector[
+                way_1 * 0.5 * (column_section_length_1 + footing.effective_height),
+                way_2 * 0.5 * (column_section_length_2 + footing.effective_height),
+                0.0
+              ]
+            end
+
+            def add_relative_location_from_a_vector(vector)
+              relative_location = Engineering::Locations::Relative.new(
+                value_1: 0, value_2: 0, value_3: 0
+              )
+              relative_location.update_from_vector(vector)
+              local_coordinates_system.add_location(relative_location)
+            end
+
+            def relative_location_at_index(index)
+              local_coordinates_system.relative_locations[index]
+            end
```

Comment:
![permeter_verties_relative-to-column](https://user-images.githubusercontent.com/26731448/230076825-2210cb6e-4df8-4121-a98e-b65f144a632c.gif)